# Data Cleaning

In [486]:
import pandas as pd
import numpy as np
import seaborn as sns


## Accessment of Data

In [487]:
# Making the copy of the dataset 

df_fifa = pd.read_csv('Fifa_world_cup3.csv')
fifa = df_fifa.copy()

### Data Summery 

- The dataset represents historical information on FIFA World Cup matches. It includes details about home and away teams, match scores, penalties, goals, and other match-specific data like the stadium, city, attendance, and referee information. There are 865 matches in the dataset, and certain columns contain incomplete data, especially for goals scored by the first and second teams, and referee cities. Some columns, such as attendance, contain non-numeric values, and date formatting is inconsistent.

### Info About Each Column

- __Home_Team:__ The name of the home team (object type).
- __Score:__ The match score (object type).
- __Away_Team:__ The name of the away team (object type).
- __Penalties:__ Indicates whether penalties were involved (object type).
- __First_Team_Goals:__ Descriptions of goals scored by the first team (object type).
- __Second_Team_Goals:__ Descriptions of goals scored by the second team (object type).
- __Date:__ Date of the match (object type).
- __Time:__ Time of the match (object type).
- __Stadium:__ Stadium where the match was played (object type).
- __City:__ City where the match was played (object type).
- __Attendance:__ The number of spectators (object type).
- __Referee:__ The referee of the match (object type).
- __Referee_City:__ The city where the referee is from (object type).

In [488]:
fifa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 929 entries, 0 to 928
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Home_Team          929 non-null    object
 1   Score              929 non-null    object
 2   Away_Team          929 non-null    object
 3   Penalties          929 non-null    object
 4   First_Team_Goals   722 non-null    object
 5   Second_Team_Goals  602 non-null    object
 6   Date               929 non-null    object
 7   Time               928 non-null    object
 8   Stadium            929 non-null    object
 9   City               907 non-null    object
 10  Attendance         928 non-null    object
 11  Referee            928 non-null    object
 12  Referee_City       887 non-null    object
dtypes: object(13)
memory usage: 94.5+ KB


In [489]:
fifa.describe()

Home_Team Score Away_Team Penalties First_Team_Goals  \
count        929   929       929       929              722   
unique        81    69        87        14              721   
top      Brazil    1–0    Mexico        NO    Houseman  20'   
freq          82   100        41       894                2   

                        Second_Team_Goals         Date   Time         Stadium  \
count                                 602          929    928             929   
unique                                602          365    125             209   
top     Mbappé 80' (pen.) 81' 118' (pen.)  27 May 1934  16:00  Estadio Azteca   
freq                                    1            8     80              19   

               City          Attendance          Referee Referee_City  
count           907                 928              928          887  
unique          169                 720              410           88  
top     Mexico City  Attendance: 45,000  Ravshan Irmatov        Italy  
freq             23                  13               11           51

In [490]:
fifa[fifa['Time'].isnull()]

Home_Team Score Away_Team Penalties First_Team_Goals Second_Team_Goals  \
37   Sweden    w/o   Austria        NO              NaN               NaN   

           Date Time        Stadium  City Attendance Referee Referee_City  
37  5 June 1938  NaN  Stade Gerland  Lyon        NaN     NaN          NaN

In [491]:
fifa['First_Team_Goals'].isnull().sum()

np.int64(207)

In [492]:
fifa['Second_Team_Goals'].isnull().sum()

np.int64(327)

In [493]:
fifa.head()

Home_Team Score Away_Team Penalties  \
0     France    4–1    Mexico        NO   
1  Argentina    1–0    France        NO   
2      Chile    3–0    Mexico        NO   
3      Chile    1–0    France        NO   
4  Argentina    6–3    Mexico        NO   

                                    First_Team_Goals  \
0  L. Laurent  19'; Langiller  40'; Maschinot  43...   
1                                         Monti  81'   
2         Vidal  3', 65'; M. Rosas  52' (o.g.); o.g.   
3                                      Subiabre  67'   
4  Stábile  8', 17', 80'; Zumelzú  12', 55'; Vara...   

                             Second_Team_Goals          Date  \
0                                 Carreño  70'  13 July 1930   
1                                          NaN  15 July 1930   
2                                          NaN  16 July 1930   
3                                          NaN  19 July 1930   
4  M. Rosas  42' (pen.), 65'; pen.  75'; Gayón  19 July 1930   

                    Time                 Stadium        City  \
0  15:00 UYT (UTC−03:30)         Estadio Pocitos  Montevideo   
1  16:00 UYT (UTC−03:30)  Estadio Parque Central  Montevideo   
2  14:45 UYT (UTC−03:30)  Estadio Parque Central  Montevideo   
3  12:50 UYT (UTC−03:30)      Estadio Centenario  Montevideo   
4  15:00 UYT (UTC−03:30)      Estadio Centenario  Montevideo   

           Attendance           Referee Referee_City  
0   Attendance: 4,444           Uruguay          NaN  
1  Attendance: 23,409      Almeida Rêgo       Brazil  
2   Attendance: 9,249  Henri Christophe      Belgium  
3   Attendance: 2,000           Uruguay          NaN  
4  Attendance: 42,100    Ulises Saucedo      Bolivia

In [494]:
fifa.tail()

Home_Team         Score Away_Team Penalties  \
924    England            1–2    France        NO   
925  Argentina            3–0   Croatia        NO   
926     France            2–0   Morocco        NO   
927    Croatia            2–1   Morocco        NO   
928  Argentina   3–3 (a.e.t.)    France       4–2   

                        First_Team_Goals                  Second_Team_Goals  \
924                      Kane 54' (pen.)         Tchouaméni 17'; Giroud 78'   
925    Messi 34' (pen.); Álvarez 39' 69'                                NaN   
926      T. Hernandez 5'; Kolo Muani 79'                                NaN   
927               Gvardiol 7'; Oršić 42'                            Dari 9'   
928  Messi 23' (pen.) 108'; Di María 36'  Mbappé 80' (pen.) 81' 118' (pen.)   

                              Date   Time                        Stadium  \
924  10 December 2022 (2022-12-10)  22:00                Al Bayt Stadium   
925  13 December 2022 (2022-12-13)  22:00                 Lusail Stadium   
926  14 December 2022 (2022-12-14)  22:00                Al Bayt Stadium   
927  17 December 2022 (2022-12-17)  18:00  Khalifa International Stadium   
928  18 December 2022 (2022-12-18)  18:00                 Lusail Stadium   

          City          Attendance                Referee Referee_City  
924    Al Khor  Attendance: 68,895         Wilton Sampaio       Brazil  
925     Lusail  Attendance: 88,966         Daniele Orsato        Italy  
926    Al Khor  Attendance: 68,294     César Arturo Ramos       Mexico  
927  Al Rayyan  Attendance: 44,137  Abdulrahman Al-Jassim        Qatar  
928     Lusail  Attendance: 88,966       Szymon Marciniak       Poland

In [495]:
fifa.sample(5)

Home_Team Score      Away_Team Penalties  \
267     Poland    0–1   West Germany        NO   
326  Argentina    2–0    El Salvador        NO   
683    Nigeria    2–2    South Korea        NO   
713   Portugal    0–0         Brazil        NO   
923    Morocco    1–0       Portugal        NO   

                               First_Team_Goals  \
267                                         NaN   
326  Passarella  22' (pen.); pen.  54'; Bertoni   
683                 Uche 12'; Yakubu 69' (pen.)   
713                                         NaN   
923                               En-Nesyri 42'   

                        Second_Team_Goals                           Date  \
267                           Müller  76'                    3 July 1974   
326                                   NaN                   23 June 1982   
683  Lee Jung-soo 38'; Park Chu-young 49'                   22 June 2010   
713                                   NaN                   25 June 2010   
923                                   NaN  10 December 2022 (2022-12-10)   

           Time                  Stadium       City          Attendance  \
267    16:35[b]              Waldstadion  Frankfurt  Attendance: 62,000   
326  21:00 CEST  Estadio José Rico Pérez   Alicante  Attendance: 32,500   
683       20:30    Moses Mabhida Stadium     Durban  Attendance: 61,874   
713       16:00    Moses Mabhida Stadium     Durban  Attendance: 62,712   
923       18:00       Al Thumama Stadium       Doha  Attendance: 44,198   

                  Referee Referee_City  
267        Erich Linemayr      Austria  
326        Luis Barrancos      Bolivia  
683  Olegário Benquerença     Portugal  
713      Benito Archundia       Mexico  
923         Facundo Tello    Argentina

### Dirty Data vs Messy Data

#### Dirty Data

- Home_Team
    1. Need to strip the spaces in the name __Consistency__
    2. Need to change all the letter into smaller cases __Consistency__

- Attendance
    1. The values contain non-numeric characters (e.g., "Attendance: 4,444").
        __Validity__

- First_Team_Goals & Second_Team_Goals
    1. The format includes textual representations of goals, such as "L. Laurent 19'" and mixed data types. __Accuracy__

- Date
    1. Inconsistent formats like "8 June 1958" and "15 July 1930". Need to remove one of the date.__Consistency__
    2. Also need to change the type to datetime. __Accuracy__

- Referee_City
    1. Missing data in several rows.
       __Completeness__

- Penalties
    1. There are only two values: "YES" and "NO". It can be simplified to binary (1 or 0). __Consistency__

- Score
    1. The scores are represented in a non-standardized format, e.g., "4–1". __Consistency__
    2. Need to change the type to int __Validity__

- Time
    1. Need to remove extra written text. __Validity__
    2. Need to change the column type datetime __Consistency__

#### Messy Data

- First_Team_Goal
    1. Need to make one more column in which it will show the player name who goal and at what time for Home team

- Second_Team_Goal 
    1. Need to make one more column in which it will show the player name who goal and at what time for away team

- Score 
    1. Need to make the two different row. One for Home teams score and away teams score team. Then one more for the winner team.


### Define, code, test

In [496]:
# Home_Team: Strip spaces and convert to lowercase
fifa['Home_Team'] = fifa['Home_Team'].str.strip().str.lower()

In [497]:
fifa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 929 entries, 0 to 928
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Home_Team          929 non-null    object
 1   Score              929 non-null    object
 2   Away_Team          929 non-null    object
 3   Penalties          929 non-null    object
 4   First_Team_Goals   722 non-null    object
 5   Second_Team_Goals  602 non-null    object
 6   Date               929 non-null    object
 7   Time               928 non-null    object
 8   Stadium            929 non-null    object
 9   City               907 non-null    object
 10  Attendance         928 non-null    object
 11  Referee            928 non-null    object
 12  Referee_City       887 non-null    object
dtypes: object(13)
memory usage: 94.5+ KB


In [498]:
fifa['First_Team_Goals'].fillna(0, inplace=True)

C:\Users\User\AppData\Local\Temp\ipykernel_13604\1134096415.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  fifa['First_Team_Goals'].fillna(0, inplace=True)


In [499]:
fifa['Second_Team_Goals'].fillna(0, inplace=True)

C:\Users\User\AppData\Local\Temp\ipykernel_13604\2102830755.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  fifa['Second_Team_Goals'].fillna(0, inplace=True)


In [500]:
fifa['City'].fillna('Unknown', inplace=True)

C:\Users\User\AppData\Local\Temp\ipykernel_13604\703393346.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  fifa['City'].fillna('Unknown', inplace=True)


In [501]:
fifa['Referee_City'].fillna('Unknown', inplace=True)

C:\Users\User\AppData\Local\Temp\ipykernel_13604\3027417422.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  fifa['Referee_City'].fillna('Unknown', inplace=True)


In [502]:
fifa.head()

Home_Team Score Away_Team Penalties  \
0     france   4–1    Mexico        NO   
1  argentina   1–0    France        NO   
2      chile   3–0    Mexico        NO   
3      chile   1–0    France        NO   
4  argentina   6–3    Mexico        NO   

                                    First_Team_Goals  \
0  L. Laurent  19'; Langiller  40'; Maschinot  43...   
1                                         Monti  81'   
2         Vidal  3', 65'; M. Rosas  52' (o.g.); o.g.   
3                                      Subiabre  67'   
4  Stábile  8', 17', 80'; Zumelzú  12', 55'; Vara...   

                             Second_Team_Goals          Date  \
0                                 Carreño  70'  13 July 1930   
1                                            0  15 July 1930   
2                                            0  16 July 1930   
3                                            0  19 July 1930   
4  M. Rosas  42' (pen.), 65'; pen.  75'; Gayón  19 July 1930   

                    Time                 Stadium        City  \
0  15:00 UYT (UTC−03:30)         Estadio Pocitos  Montevideo   
1  16:00 UYT (UTC−03:30)  Estadio Parque Central  Montevideo   
2  14:45 UYT (UTC−03:30)  Estadio Parque Central  Montevideo   
3  12:50 UYT (UTC−03:30)      Estadio Centenario  Montevideo   
4  15:00 UYT (UTC−03:30)      Estadio Centenario  Montevideo   

           Attendance           Referee Referee_City  
0   Attendance: 4,444           Uruguay      Unknown  
1  Attendance: 23,409      Almeida Rêgo       Brazil  
2   Attendance: 9,249  Henri Christophe      Belgium  
3   Attendance: 2,000           Uruguay      Unknown  
4  Attendance: 42,100    Ulises Saucedo      Bolivia

In [503]:
import re
def clean_attendance(value):
    if pd.isnull(value):  
        return None
    numeric_value = re.sub(r'[^0-9]', '', str(value))
    return int(numeric_value) if numeric_value else None

fifa['Attendance'] = fifa['Attendance'].apply(clean_attendance)


In [504]:
fifa.head()

Home_Team Score Away_Team Penalties  \
0     france   4–1    Mexico        NO   
1  argentina   1–0    France        NO   
2      chile   3–0    Mexico        NO   
3      chile   1–0    France        NO   
4  argentina   6–3    Mexico        NO   

                                    First_Team_Goals  \
0  L. Laurent  19'; Langiller  40'; Maschinot  43...   
1                                         Monti  81'   
2         Vidal  3', 65'; M. Rosas  52' (o.g.); o.g.   
3                                      Subiabre  67'   
4  Stábile  8', 17', 80'; Zumelzú  12', 55'; Vara...   

                             Second_Team_Goals          Date  \
0                                 Carreño  70'  13 July 1930   
1                                            0  15 July 1930   
2                                            0  16 July 1930   
3                                            0  19 July 1930   
4  M. Rosas  42' (pen.), 65'; pen.  75'; Gayón  19 July 1930   

                    Time                 Stadium        City  Attendance  \
0  15:00 UYT (UTC−03:30)         Estadio Pocitos  Montevideo      4444.0   
1  16:00 UYT (UTC−03:30)  Estadio Parque Central  Montevideo     23409.0   
2  14:45 UYT (UTC−03:30)  Estadio Parque Central  Montevideo      9249.0   
3  12:50 UYT (UTC−03:30)      Estadio Centenario  Montevideo      2000.0   
4  15:00 UYT (UTC−03:30)      Estadio Centenario  Montevideo     42100.0   

            Referee Referee_City  
0           Uruguay      Unknown  
1      Almeida Rêgo       Brazil  
2  Henri Christophe      Belgium  
3           Uruguay      Unknown  
4    Ulises Saucedo      Bolivia

In [505]:
fifa.tail()

Home_Team         Score Away_Team Penalties  \
924    england           1–2    France        NO   
925  argentina           3–0   Croatia        NO   
926     france           2–0   Morocco        NO   
927    croatia           2–1   Morocco        NO   
928  argentina  3–3 (a.e.t.)    France       4–2   

                        First_Team_Goals                  Second_Team_Goals  \
924                      Kane 54' (pen.)         Tchouaméni 17'; Giroud 78'   
925    Messi 34' (pen.); Álvarez 39' 69'                                  0   
926      T. Hernandez 5'; Kolo Muani 79'                                  0   
927               Gvardiol 7'; Oršić 42'                            Dari 9'   
928  Messi 23' (pen.) 108'; Di María 36'  Mbappé 80' (pen.) 81' 118' (pen.)   

                              Date   Time                        Stadium  \
924  10 December 2022 (2022-12-10)  22:00                Al Bayt Stadium   
925  13 December 2022 (2022-12-13)  22:00                 Lusail Stadium   
926  14 December 2022 (2022-12-14)  22:00                Al Bayt Stadium   
927  17 December 2022 (2022-12-17)  18:00  Khalifa International Stadium   
928  18 December 2022 (2022-12-18)  18:00                 Lusail Stadium   

          City  Attendance                Referee Referee_City  
924    Al Khor     68895.0         Wilton Sampaio       Brazil  
925     Lusail     88966.0         Daniele Orsato        Italy  
926    Al Khor     68294.0     César Arturo Ramos       Mexico  
927  Al Rayyan     44137.0  Abdulrahman Al-Jassim        Qatar  
928     Lusail     88966.0       Szymon Marciniak       Poland

In [506]:
import re

# Function to clean attendance values by removing non-numeric characters
def clean_attendance(value):
    numeric_value = re.sub(r'[^0-9]', '', str(value))
    return int(numeric_value) if numeric_value else None

fifa['Attendance'] = fifa['Attendance'].apply(clean_attendance)

In [507]:
fifa.head()

Home_Team Score Away_Team Penalties  \
0     france   4–1    Mexico        NO   
1  argentina   1–0    France        NO   
2      chile   3–0    Mexico        NO   
3      chile   1–0    France        NO   
4  argentina   6–3    Mexico        NO   

                                    First_Team_Goals  \
0  L. Laurent  19'; Langiller  40'; Maschinot  43...   
1                                         Monti  81'   
2         Vidal  3', 65'; M. Rosas  52' (o.g.); o.g.   
3                                      Subiabre  67'   
4  Stábile  8', 17', 80'; Zumelzú  12', 55'; Vara...   

                             Second_Team_Goals          Date  \
0                                 Carreño  70'  13 July 1930   
1                                            0  15 July 1930   
2                                            0  16 July 1930   
3                                            0  19 July 1930   
4  M. Rosas  42' (pen.), 65'; pen.  75'; Gayón  19 July 1930   

                    Time                 Stadium        City  Attendance  \
0  15:00 UYT (UTC−03:30)         Estadio Pocitos  Montevideo     44440.0   
1  16:00 UYT (UTC−03:30)  Estadio Parque Central  Montevideo    234090.0   
2  14:45 UYT (UTC−03:30)  Estadio Parque Central  Montevideo     92490.0   
3  12:50 UYT (UTC−03:30)      Estadio Centenario  Montevideo     20000.0   
4  15:00 UYT (UTC−03:30)      Estadio Centenario  Montevideo    421000.0   

            Referee Referee_City  
0           Uruguay      Unknown  
1      Almeida Rêgo       Brazil  
2  Henri Christophe      Belgium  
3           Uruguay      Unknown  
4    Ulises Saucedo      Bolivia

In [508]:
fifa['Date'] = fifa['Date'].str.strip().str.split("(" ).str.get(0)

In [509]:
fifa['Date'] = pd.DatetimeIndex(fifa['Date'])

In [510]:
fifa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 929 entries, 0 to 928
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Home_Team          929 non-null    object        
 1   Score              929 non-null    object        
 2   Away_Team          929 non-null    object        
 3   Penalties          929 non-null    object        
 4   First_Team_Goals   929 non-null    object        
 5   Second_Team_Goals  929 non-null    object        
 6   Date               929 non-null    datetime64[ns]
 7   Time               928 non-null    object        
 8   Stadium            929 non-null    object        
 9   City               929 non-null    object        
 10  Attendance         928 non-null    float64       
 11  Referee            928 non-null    object        
 12  Referee_City       929 non-null    object        
dtypes: datetime64[ns](1), float64(1), object(11)
memory usage: 94.5+ 

In [511]:
fifa.tail()

Home_Team         Score Away_Team Penalties  \
924    england           1–2    France        NO   
925  argentina           3–0   Croatia        NO   
926     france           2–0   Morocco        NO   
927    croatia           2–1   Morocco        NO   
928  argentina  3–3 (a.e.t.)    France       4–2   

                        First_Team_Goals                  Second_Team_Goals  \
924                      Kane 54' (pen.)         Tchouaméni 17'; Giroud 78'   
925    Messi 34' (pen.); Álvarez 39' 69'                                  0   
926      T. Hernandez 5'; Kolo Muani 79'                                  0   
927               Gvardiol 7'; Oršić 42'                            Dari 9'   
928  Messi 23' (pen.) 108'; Di María 36'  Mbappé 80' (pen.) 81' 118' (pen.)   

          Date   Time                        Stadium       City  Attendance  \
924 2022-12-10  22:00                Al Bayt Stadium    Al Khor    688950.0   
925 2022-12-13  22:00                 Lusail Stadium     Lusail    889660.0   
926 2022-12-14  22:00                Al Bayt Stadium    Al Khor    682940.0   
927 2022-12-17  18:00  Khalifa International Stadium  Al Rayyan    441370.0   
928 2022-12-18  18:00                 Lusail Stadium     Lusail    889660.0   

                   Referee Referee_City  
924         Wilton Sampaio       Brazil  
925         Daniele Orsato        Italy  
926     César Arturo Ramos       Mexico  
927  Abdulrahman Al-Jassim        Qatar  
928       Szymon Marciniak       Poland

In [512]:
fifa.head()

Home_Team Score Away_Team Penalties  \
0     france   4–1    Mexico        NO   
1  argentina   1–0    France        NO   
2      chile   3–0    Mexico        NO   
3      chile   1–0    France        NO   
4  argentina   6–3    Mexico        NO   

                                    First_Team_Goals  \
0  L. Laurent  19'; Langiller  40'; Maschinot  43...   
1                                         Monti  81'   
2         Vidal  3', 65'; M. Rosas  52' (o.g.); o.g.   
3                                      Subiabre  67'   
4  Stábile  8', 17', 80'; Zumelzú  12', 55'; Vara...   

                             Second_Team_Goals       Date  \
0                                 Carreño  70' 1930-07-13   
1                                            0 1930-07-15   
2                                            0 1930-07-16   
3                                            0 1930-07-19   
4  M. Rosas  42' (pen.), 65'; pen.  75'; Gayón 1930-07-19   

                    Time                 Stadium        City  Attendance  \
0  15:00 UYT (UTC−03:30)         Estadio Pocitos  Montevideo     44440.0   
1  16:00 UYT (UTC−03:30)  Estadio Parque Central  Montevideo    234090.0   
2  14:45 UYT (UTC−03:30)  Estadio Parque Central  Montevideo     92490.0   
3  12:50 UYT (UTC−03:30)      Estadio Centenario  Montevideo     20000.0   
4  15:00 UYT (UTC−03:30)      Estadio Centenario  Montevideo    421000.0   

            Referee Referee_City  
0           Uruguay      Unknown  
1      Almeida Rêgo       Brazil  
2  Henri Christophe      Belgium  
3           Uruguay      Unknown  
4    Ulises Saucedo      Bolivia

In [513]:
fifa['Time'] = fifa['Time'].astype('string')

In [514]:
fifa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 929 entries, 0 to 928
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Home_Team          929 non-null    object        
 1   Score              929 non-null    object        
 2   Away_Team          929 non-null    object        
 3   Penalties          929 non-null    object        
 4   First_Team_Goals   929 non-null    object        
 5   Second_Team_Goals  929 non-null    object        
 6   Date               929 non-null    datetime64[ns]
 7   Time               928 non-null    string        
 8   Stadium            929 non-null    object        
 9   City               929 non-null    object        
 10  Attendance         928 non-null    float64       
 11  Referee            928 non-null    object        
 12  Referee_City       929 non-null    object        
dtypes: datetime64[ns](1), float64(1), object(10), string(1)
memory us

In [515]:
fifa['Time'].astype('string')

0      15:00 UYT (UTC−03:30)
1      16:00 UYT (UTC−03:30)
2      14:45 UYT (UTC−03:30)
3      12:50 UYT (UTC−03:30)
4      15:00 UYT (UTC−03:30)
               ...          
924                    22:00
925                    22:00
926                    22:00
927                    18:00
928                    18:00
Name: Time, Length: 929, dtype: string

In [516]:
fifa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 929 entries, 0 to 928
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Home_Team          929 non-null    object        
 1   Score              929 non-null    object        
 2   Away_Team          929 non-null    object        
 3   Penalties          929 non-null    object        
 4   First_Team_Goals   929 non-null    object        
 5   Second_Team_Goals  929 non-null    object        
 6   Date               929 non-null    datetime64[ns]
 7   Time               928 non-null    string        
 8   Stadium            929 non-null    object        
 9   City               929 non-null    object        
 10  Attendance         928 non-null    float64       
 11  Referee            928 non-null    object        
 12  Referee_City       929 non-null    object        
dtypes: datetime64[ns](1), float64(1), object(10), string(1)
memory us

In [517]:
fifa['Time'] = fifa['Time'].str.split(" ").str.get(0)

In [518]:
def clean_time(value):
    if pd.isnull(value):  
        return np.nan
   
    time_part = value.split(' ')[0]
    return time_part 

fifa['Start_Time'] = fifa['Time'].apply(clean_time)

fifa['Start_Time'] = pd.to_datetime(fifa['Start_Time'], format='%H:%M', errors='coerce').dt.time


In [519]:
fifa['Time'] = fifa['Start_Time']

In [520]:
fifa.drop(columns=['Start_Time' ], inplace=True)

In [521]:
fifa.head()

Home_Team Score Away_Team Penalties  \
0     france   4–1    Mexico        NO   
1  argentina   1–0    France        NO   
2      chile   3–0    Mexico        NO   
3      chile   1–0    France        NO   
4  argentina   6–3    Mexico        NO   

                                    First_Team_Goals  \
0  L. Laurent  19'; Langiller  40'; Maschinot  43...   
1                                         Monti  81'   
2         Vidal  3', 65'; M. Rosas  52' (o.g.); o.g.   
3                                      Subiabre  67'   
4  Stábile  8', 17', 80'; Zumelzú  12', 55'; Vara...   

                             Second_Team_Goals       Date      Time  \
0                                 Carreño  70' 1930-07-13  15:00:00   
1                                            0 1930-07-15  16:00:00   
2                                            0 1930-07-16  14:45:00   
3                                            0 1930-07-19  12:50:00   
4  M. Rosas  42' (pen.), 65'; pen.  75'; Gayón 1930-07-19  15:00:00   

                  Stadium        City  Attendance           Referee  \
0         Estadio Pocitos  Montevideo     44440.0           Uruguay   
1  Estadio Parque Central  Montevideo    234090.0      Almeida Rêgo   
2  Estadio Parque Central  Montevideo     92490.0  Henri Christophe   
3      Estadio Centenario  Montevideo     20000.0           Uruguay   
4      Estadio Centenario  Montevideo    421000.0    Ulises Saucedo   

  Referee_City  
0      Unknown  
1       Brazil  
2      Belgium  
3      Unknown  
4      Bolivia

In [522]:
import re

# Function to extract goals, clean 'pen.' and handle multiple goal times for a player
def extract_goals(goal_column):
    player_names = []
    goal_times = []
    
    for goals in goal_column:
        # Check if the entry is not a string (to handle integers or NaN values)
        if pd.isna(goals) or not isinstance(goals, str):
            continue
        
        # Remove unwanted substrings like (pen.) and extra spaces
        cleaned_goals = goals.replace('(pen.)', '').strip()
        
        # Split the goals based on semicolon
        goal_entries = cleaned_goals.split(';')
        
        for entry in goal_entries:
            entry = entry.strip()
            # Extract the player name and associated goal times
            match = re.match(r'(.+?)\s([\d\s\']+)', entry)
            if match:
                player_name = match.group(1).strip()
                times = match.group(2).strip()
                
                # Split the times by space and repeat player name for each goal time
                for time in times.split():
                    goal_time = time.replace("'", "").strip()
                    player_names.append(player_name)
                    goal_times.append(goal_time)
                    
    return player_names, goal_times

# Create an empty list to store the expanded rows
expanded_rows = []

# Iterate through each row of the dataframe
for index, row in fifa.iterrows():
    # Extract goals and times for both first and second teams
    first_team_players, first_team_times = extract_goals([row['First_Team_Goals']])
    second_team_players, second_team_times = extract_goals([row['Second_Team_Goals']])
    
    # Combine players and times from both teams
    all_players = first_team_players + second_team_players
    all_goal_times = first_team_times + second_team_times
    
    # Create a new row for each player-goal combination
    for player, goal_time in zip(all_players, all_goal_times):
        new_row = row.copy()
        new_row['Player_Name'] = player
        new_row['Goal_Time'] = goal_time
        expanded_rows.append(new_row)

# Convert the expanded rows into a new dataframe
expanded_fifa = pd.DataFrame(expanded_rows)

# Display the expanded dataframe
expanded_fifa.head()


Home_Team Score Away_Team Penalties  \
0     france   4–1    Mexico        NO   
0     france   4–1    Mexico        NO   
0     france   4–1    Mexico        NO   
0     france   4–1    Mexico        NO   
1  argentina   1–0    France        NO   

                                    First_Team_Goals Second_Team_Goals  \
0  L. Laurent  19'; Langiller  40'; Maschinot  43...      Carreño  70'   
0  L. Laurent  19'; Langiller  40'; Maschinot  43...      Carreño  70'   
0  L. Laurent  19'; Langiller  40'; Maschinot  43...      Carreño  70'   
0  L. Laurent  19'; Langiller  40'; Maschinot  43...      Carreño  70'   
1                                         Monti  81'                 0   

        Date      Time                 Stadium        City  Attendance  \
0 1930-07-13  15:00:00         Estadio Pocitos  Montevideo     44440.0   
0 1930-07-13  15:00:00         Estadio Pocitos  Montevideo     44440.0   
0 1930-07-13  15:00:00         Estadio Pocitos  Montevideo     44440.0   
0 1930-07-13  15:00:00         Estadio Pocitos  Montevideo     44440.0   
1 1930-07-15  16:00:00  Estadio Parque Central  Montevideo    234090.0   

        Referee Referee_City Player_Name Goal_Time  
0       Uruguay      Unknown  L. Laurent        19  
0       Uruguay      Unknown   Langiller        40  
0       Uruguay      Unknown   Maschinot        43  
0       Uruguay      Unknown     Carreño        70  
1  Almeida Rêgo       Brazil       Monti        81

In [523]:
expanded_fifa.head()

Home_Team Score Away_Team Penalties  \
0     france   4–1    Mexico        NO   
0     france   4–1    Mexico        NO   
0     france   4–1    Mexico        NO   
0     france   4–1    Mexico        NO   
1  argentina   1–0    France        NO   

                                    First_Team_Goals Second_Team_Goals  \
0  L. Laurent  19'; Langiller  40'; Maschinot  43...      Carreño  70'   
0  L. Laurent  19'; Langiller  40'; Maschinot  43...      Carreño  70'   
0  L. Laurent  19'; Langiller  40'; Maschinot  43...      Carreño  70'   
0  L. Laurent  19'; Langiller  40'; Maschinot  43...      Carreño  70'   
1                                         Monti  81'                 0   

        Date      Time                 Stadium        City  Attendance  \
0 1930-07-13  15:00:00         Estadio Pocitos  Montevideo     44440.0   
0 1930-07-13  15:00:00         Estadio Pocitos  Montevideo     44440.0   
0 1930-07-13  15:00:00         Estadio Pocitos  Montevideo     44440.0   
0 1930-07-13  15:00:00         Estadio Pocitos  Montevideo     44440.0   
1 1930-07-15  16:00:00  Estadio Parque Central  Montevideo    234090.0   

        Referee Referee_City Player_Name Goal_Time  
0       Uruguay      Unknown  L. Laurent        19  
0       Uruguay      Unknown   Langiller        40  
0       Uruguay      Unknown   Maschinot        43  
0       Uruguay      Unknown     Carreño        70  
1  Almeida Rêgo       Brazil       Monti        81

In [524]:
expanded_fifa.tail(10)

Home_Team         Score Away_Team Penalties  \
926     france           2–0   Morocco        NO   
927    croatia           2–1   Morocco        NO   
927    croatia           2–1   Morocco        NO   
927    croatia           2–1   Morocco        NO   
928  argentina  3–3 (a.e.t.)    France       4–2   
928  argentina  3–3 (a.e.t.)    France       4–2   
928  argentina  3–3 (a.e.t.)    France       4–2   
928  argentina  3–3 (a.e.t.)    France       4–2   
928  argentina  3–3 (a.e.t.)    France       4–2   
928  argentina  3–3 (a.e.t.)    France       4–2   

                        First_Team_Goals                  Second_Team_Goals  \
926      T. Hernandez 5'; Kolo Muani 79'                                  0   
927               Gvardiol 7'; Oršić 42'                            Dari 9'   
927               Gvardiol 7'; Oršić 42'                            Dari 9'   
927               Gvardiol 7'; Oršić 42'                            Dari 9'   
928  Messi 23' (pen.) 108'; Di María 36'  Mbappé 80' (pen.) 81' 118' (pen.)   
928  Messi 23' (pen.) 108'; Di María 36'  Mbappé 80' (pen.) 81' 118' (pen.)   
928  Messi 23' (pen.) 108'; Di María 36'  Mbappé 80' (pen.) 81' 118' (pen.)   
928  Messi 23' (pen.) 108'; Di María 36'  Mbappé 80' (pen.) 81' 118' (pen.)   
928  Messi 23' (pen.) 108'; Di María 36'  Mbappé 80' (pen.) 81' 118' (pen.)   
928  Messi 23' (pen.) 108'; Di María 36'  Mbappé 80' (pen.) 81' 118' (pen.)   

          Date      Time                        Stadium       City  \
926 2022-12-14  22:00:00                Al Bayt Stadium    Al Khor   
927 2022-12-17  18:00:00  Khalifa International Stadium  Al Rayyan   
927 2022-12-17  18:00:00  Khalifa International Stadium  Al Rayyan   
927 2022-12-17  18:00:00  Khalifa International Stadium  Al Rayyan   
928 2022-12-18  18:00:00                 Lusail Stadium     Lusail   
928 2022-12-18  18:00:00                 Lusail Stadium     Lusail   
928 2022-12-18  18:00:00                 Lusail Stadium     Lusail   
928 2022-12-18  18:00:00                 Lusail Stadium     Lusail   
928 2022-12-18  18:00:00                 Lusail Stadium     Lusail   
928 2022-12-18  18:00:00                 Lusail Stadium     Lusail   

     Attendance                Referee Referee_City Player_Name Goal_Time  
926    682940.0     César Arturo Ramos       Mexico  Kolo Muani        79  
927    441370.0  Abdulrahman Al-Jassim        Qatar    Gvardiol         7  
927    441370.0  Abdulrahman Al-Jassim        Qatar       Oršić        42  
927    441370.0  Abdulrahman Al-Jassim        Qatar        Dari         9  
928    889660.0       Szymon Marciniak       Poland       Messi        23  
928    889660.0       Szymon Marciniak       Poland       Messi       108  
928    889660.0       Szymon Marciniak       Poland    Di María        36  
928    889660.0       Szymon Marciniak       Poland      Mbappé        80  
928    889660.0       Szymon Marciniak       Poland      Mbappé        81  
928    889660.0       Szymon Marciniak       Poland      Mbappé       118

In [525]:
df = expanded_fifa.copy()

In [526]:
df.reset_index(drop=True, inplace=True)


In [536]:
temp_df = df[df['Referee_City'] == 'Unknown']
temp_df

Home_Team Score          Away_Team Penalties  \
0          france   4–1             Mexico        NO   
1          france   4–1             Mexico        NO   
2          france   4–1             Mexico        NO   
3          france   4–1             Mexico        NO   
7           chile   1–0             France        NO   
..            ...   ...                ...       ...   
423        brazil   1–0              Wales        NO   
424        france   4–0   Northern Ireland        NO   
425        france   4–0   Northern Ireland        NO   
426        france   4–0   Northern Ireland        NO   
429  west germany   1–0         Yugoslavia        NO   

                                      First_Team_Goals Second_Team_Goals  \
0    L. Laurent  19'; Langiller  40'; Maschinot  43...      Carreño  70'   
1    L. Laurent  19'; Langiller  40'; Maschinot  43...      Carreño  70'   
2    L. Laurent  19'; Langiller  40'; Maschinot  43...      Carreño  70'   
3    L. Laurent  19'; Langiller  40'; Maschinot  43...      Carreño  70'   
7                                        Subiabre  67'                 0   
..                                                 ...               ...   
423                                          Pelé  66'                 0   
424  Wisnieski  44'; Fontaine  55', 63'; Piantoni  68'                 0   
425  Wisnieski  44'; Fontaine  55', 63'; Piantoni  68'                 0   
426  Wisnieski  44'; Fontaine  55', 63'; Piantoni  68'                 0   
429                                          Rahn  12'                 0   

          Date      Time             Stadium        City  Attendance  \
0   1930-07-13  15:00:00     Estadio Pocitos  Montevideo     44440.0   
1   1930-07-13  15:00:00     Estadio Pocitos  Montevideo     44440.0   
2   1930-07-13  15:00:00     Estadio Pocitos  Montevideo     44440.0   
3   1930-07-13  15:00:00     Estadio Pocitos  Montevideo     44440.0   
7   1930-07-19  12:50:00  Estadio Centenario  Montevideo     20000.0   
..         ...       ...                 ...         ...         ...   
423 1958-06-19  19:00:00              Ullevi  Gothenburg    259230.0   
424 1958-06-19  19:00:00       Idrottsparken  Norrköping    118000.0   
425 1958-06-19  19:00:00       Idrottsparken  Norrköping    118000.0   
426 1958-06-19  19:00:00       Idrottsparken  Norrköping    118000.0   
429 1958-06-19  19:00:00       Malmö Stadion       Malmö    200550.0   

         Referee Referee_City Player_Name Goal_Time  Home_Score  Away_Score  \
0        Uruguay      Unknown  L. Laurent        19         4.0         1.0   
1        Uruguay      Unknown   Langiller        40         4.0         1.0   
2        Uruguay      Unknown   Maschinot        43         4.0         1.0   
3        Uruguay      Unknown     Carreño        70         4.0         1.0   
7        Uruguay      Unknown    Subiabre        67         1.0         0.0   
..           ...          ...         ...       ...         ...         ...   
423      Austria      Unknown        Pelé        66         1.0         0.0   
424        Spain      Unknown   Wisnieski        44         4.0         0.0   
425        Spain      Unknown    Fontaine        55         4.0         0.0   
426        Spain      Unknown    Piantoni        68         4.0         0.0   
429  Switzerland      Unknown        Rahn        12         1.0         0.0   

      Outcome  
0    Home Win  
1    Home Win  
2    Home Win  
3    Home Win  
7    Home Win  
..        ...  
423  Home Win  
424  Home Win  
425  Home Win  
426  Home Win  
429  Home Win  

[131 rows x 18 columns]

In [537]:
df.loc[temp_df.index, "Referee_City"] = temp_df['Referee']


In [538]:
df.loc[temp_df.index, "Referee"] = 'unknown'


In [539]:
df.head()

Home_Team Score Away_Team Penalties  \
0     france   4–1    Mexico        NO   
1     france   4–1    Mexico        NO   
2     france   4–1    Mexico        NO   
3     france   4–1    Mexico        NO   
4  argentina   1–0    France        NO   

                                    First_Team_Goals Second_Team_Goals  \
0  L. Laurent  19'; Langiller  40'; Maschinot  43...      Carreño  70'   
1  L. Laurent  19'; Langiller  40'; Maschinot  43...      Carreño  70'   
2  L. Laurent  19'; Langiller  40'; Maschinot  43...      Carreño  70'   
3  L. Laurent  19'; Langiller  40'; Maschinot  43...      Carreño  70'   
4                                         Monti  81'                 0   

        Date      Time                 Stadium        City  Attendance  \
0 1930-07-13  15:00:00         Estadio Pocitos  Montevideo     44440.0   
1 1930-07-13  15:00:00         Estadio Pocitos  Montevideo     44440.0   
2 1930-07-13  15:00:00         Estadio Pocitos  Montevideo     44440.0   
3 1930-07-13  15:00:00         Estadio Pocitos  Montevideo     44440.0   
4 1930-07-15  16:00:00  Estadio Parque Central  Montevideo    234090.0   

        Referee Referee_City Player_Name Goal_Time  Home_Score  Away_Score  \
0       unknown      Uruguay  L. Laurent        19         4.0         1.0   
1       unknown      Uruguay   Langiller        40         4.0         1.0   
2       unknown      Uruguay   Maschinot        43         4.0         1.0   
3       unknown      Uruguay     Carreño        70         4.0         1.0   
4  Almeida Rêgo       Brazil       Monti        81         1.0         0.0   

    Outcome  
0  Home Win  
1  Home Win  
2  Home Win  
3  Home Win  
4  Home Win

In [540]:
df['Home_Score'] = df['Score'].str.extract(r'(\d+)').astype(float)
df['Away_Score'] = df['Score'].str.extract(r'–(\d+)').astype(float)

# Create a column for match outcome: Home Win, Away Win, or Draw
df['Outcome'] = df.apply(lambda row: 'Home Win' if row['Home_Score'] > row['Away_Score'] 
                                   else ('Away Win' if row['Away_Score'] > row['Home_Score'] 
                                         else 'Draw'), axis=1)

In [541]:
df.head()

Home_Team Score Away_Team Penalties  \
0     france   4–1    Mexico        NO   
1     france   4–1    Mexico        NO   
2     france   4–1    Mexico        NO   
3     france   4–1    Mexico        NO   
4  argentina   1–0    France        NO   

                                    First_Team_Goals Second_Team_Goals  \
0  L. Laurent  19'; Langiller  40'; Maschinot  43...      Carreño  70'   
1  L. Laurent  19'; Langiller  40'; Maschinot  43...      Carreño  70'   
2  L. Laurent  19'; Langiller  40'; Maschinot  43...      Carreño  70'   
3  L. Laurent  19'; Langiller  40'; Maschinot  43...      Carreño  70'   
4                                         Monti  81'                 0   

        Date      Time                 Stadium        City  Attendance  \
0 1930-07-13  15:00:00         Estadio Pocitos  Montevideo     44440.0   
1 1930-07-13  15:00:00         Estadio Pocitos  Montevideo     44440.0   
2 1930-07-13  15:00:00         Estadio Pocitos  Montevideo     44440.0   
3 1930-07-13  15:00:00         Estadio Pocitos  Montevideo     44440.0   
4 1930-07-15  16:00:00  Estadio Parque Central  Montevideo    234090.0   

        Referee Referee_City Player_Name Goal_Time  Home_Score  Away_Score  \
0       unknown      Uruguay  L. Laurent        19         4.0         1.0   
1       unknown      Uruguay   Langiller        40         4.0         1.0   
2       unknown      Uruguay   Maschinot        43         4.0         1.0   
3       unknown      Uruguay     Carreño        70         4.0         1.0   
4  Almeida Rêgo       Brazil       Monti        81         1.0         0.0   

    Outcome  
0  Home Win  
1  Home Win  
2  Home Win  
3  Home Win  
4  Home Win

In [542]:
new_order = ['Home_Team', 'Away_Team' , 'Home_Score' , 'Away_Score', 'Penalties', 'Player_Name', 'Goal_Time', 'Outcome' , 'Date', 'Time', 'Stadium', 'City', 'Attendance', 'Referee', 'Referee_City']

fifa_data_reordered = df[new_order]

fifa_data_reordered.head()

Home_Team Away_Team  Home_Score  Away_Score Penalties Player_Name  \
0     france    Mexico         4.0         1.0        NO  L. Laurent   
1     france    Mexico         4.0         1.0        NO   Langiller   
2     france    Mexico         4.0         1.0        NO   Maschinot   
3     france    Mexico         4.0         1.0        NO     Carreño   
4  argentina    France         1.0         0.0        NO       Monti   

  Goal_Time   Outcome       Date      Time                 Stadium  \
0        19  Home Win 1930-07-13  15:00:00         Estadio Pocitos   
1        40  Home Win 1930-07-13  15:00:00         Estadio Pocitos   
2        43  Home Win 1930-07-13  15:00:00         Estadio Pocitos   
3        70  Home Win 1930-07-13  15:00:00         Estadio Pocitos   
4        81  Home Win 1930-07-15  16:00:00  Estadio Parque Central   

         City  Attendance       Referee Referee_City  
0  Montevideo     44440.0       unknown      Uruguay  
1  Montevideo     44440.0       unknown      Uruguay  
2  Montevideo     44440.0       unknown      Uruguay  
3  Montevideo     44440.0       unknown      Uruguay  
4  Montevideo    234090.0  Almeida Rêgo       Brazil

In [545]:
fifa_data_reordered.tail()

Home_Team Away_Team  Home_Score  Away_Score Penalties Player_Name  \
2337  argentina    France         3.0         3.0       4–2       Messi   
2338  argentina    France         3.0         3.0       4–2    Di María   
2339  argentina    France         3.0         3.0       4–2      Mbappé   
2340  argentina    France         3.0         3.0       4–2      Mbappé   
2341  argentina    France         3.0         3.0       4–2      Mbappé   

     Goal_Time Outcome       Date      Time         Stadium    City  \
2337       108    Draw 2022-12-18  18:00:00  Lusail Stadium  Lusail   
2338        36    Draw 2022-12-18  18:00:00  Lusail Stadium  Lusail   
2339        80    Draw 2022-12-18  18:00:00  Lusail Stadium  Lusail   
2340        81    Draw 2022-12-18  18:00:00  Lusail Stadium  Lusail   
2341       118    Draw 2022-12-18  18:00:00  Lusail Stadium  Lusail   

      Attendance           Referee Referee_City  
2337    889660.0  Szymon Marciniak       Poland  
2338    889660.0  Szymon Marciniak       Poland  
2339    889660.0  Szymon Marciniak       Poland  
2340    889660.0  Szymon Marciniak       Poland  
2341    889660.0  Szymon Marciniak       Poland

In [544]:
fifa_data_reordered.to_csv('Fifa_world_cup_5.csv', index=False)